pkg/fieldpath/merge.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldpath
+
+import (
+	"reflect"
+
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+)
+
+const (
+	errInvalidMerge = "failed to merge values"
+)
+
+// MergeValue of the receiver p at the specified field path with the supplied
+// value according to supplied merge options
+func (p *Paved) MergeValue(path string, value interface{}, mo *xpv1.MergeOptions) error {
+	dst, err := p.GetValue(path)
+	if IsNotFound(err) || mo == nil {
+		dst = nil
+	} else if err != nil {
+		return err
+	}
+
+	dst, err = merge(dst, value, mo)
+	if err != nil {
+		return err
+	}
+
+	return p.SetValue(path, dst)
+}
+
+// merges the given src onto the given dst.
+// dst and src must have the same map type.
+// If a nil merge options is supplied, the default behavior is MergeOptions'
+// default behavior. If dst or src is nil, src is returned
+// (i.e., dst replaced by src).
+func merge(dst, src interface{}, mergeOptions *xpv1.MergeOptions) (interface{}, error) {
+	const keyArg = "arg"
+	argWrap := func(arg interface{}) map[string]interface{} {
+		return map[string]interface{}{
+			keyArg: arg,
+		}
+	}
```

Comment:
Could you add a comment explaining why this `argWrap` function is needed? I can't quite figure it out.

Reply:
Done.